src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -741,13 +741,32 @@ class NBDWatchCtx : public librbd::UpdateWatchCtx
       int ret;
 
       if (new_size != size) {
+        unsigned long _size = 0;
+
         dout(5) << "resize detected" << dendl;
         if (ioctl(fd, BLKFLSBUF, NULL) < 0)
           derr << "invalidate page cache failed: " << cpp_strerror(errno)
                << dendl;
-	if (use_netlink) {
+	if (!use_ioctl) {
 	  ret = netlink_resize(nbd_index, new_size);
-	} else {
+
+	  if (!ret) {
+            std::string nbd_path = "/sys/block/nbd" + stringify(nbd_index);
+            if(access(nbd_path.c_str(), F_OK) == 0) {
+              std::ifstream ifs;
+              ifs.open(nbd_path + "/size", std::ifstream::in);
+              if (ifs.is_open())
+                ifs >> _size;
```

Comment:
Shouldn't you multiply the result by RBD_NBD_BLKSIZE?

And may be `check_device_size` function could be tweaked a little and reused here? I suppose it would need to be modified to print the error message to a stream provided as an argument and add flag to skip `if (size == 0)` clause.

Alternatively a new static function `get_device_size` could be provided and used here and in `check_device_size`.

Reply:
> Shouldn't you multiply the result by RBD_NBD_BLKSIZE?
> 
Right.

> And may be `check_device_size` function could be tweaked a little and reused here? I suppose it would need to be modified to print the error message to a stream provided as an argument and add flag to skip `if (size == 0)` clause.
> 
> Alternatively a new static function `get_device_size` could be provided and used here and in `check_device_size`.

Yeah, make sense and I added one `get_device_size()` helper.